Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp

```diff
@@ -237,8 +237,8 @@ struct JPEGLoadingContext {
 
     State state { State::NotDecoded };
 
-    u32 luma_table[64] = { 0 };
-    u32 chroma_table[64] = { 0 };
+    HashMap<u8, Array<u16, 64>> quantization_tables {};
```

Comment:
Having a HashMap for indices 0-3 seems a bit overkill. Can't this be an array of arrays?

(If the motivation is if a table is accessed that's not initialized: Maybe an array of optional arrays, but I'm guessing the spec defines some initial value for the data in these arrays?)

Reply:
An HashMap is the theoretical perfect fit IMO, but obviously overkill for 4 values. We could replace it with an array and do the job manually, but I would rather go the "mathematical" way for now and replace it later when the decoder will be optimized enough to see a difference.

And no, the JPEG spec doesn't define any default value:
`If a table has never been defined for a particular destination, then when this destination is specified in a frame header, the results are unpredictable.`

Reply:
IMHO, if anything has integers 0-n with smallish n as key, then an array is the perfect fit for anyone with just a modicum of "silicon sympathy" (the usual term is "mechanical sympathy", which used to be about mechanics and actual engineering and which software folks coopted. I can't find a good link to describe it, https://dzone.com/articles/mechanical-sympathy is alright I suppose).

A hash table is also just an array, except it does a bunch of math and key reordering, so it's a useless layer of indirection in this case.

The spec says "default value can be whatever", so zeroes is fine :)

In the end, up to you, but imho using hashtables when the keys are ints 0-n with smallish n is pretty gross :)

Reply:
Re parens: 

Looks like it was popularized [here](https://www.se-radio.net/2014/02/episode-201-martin-thompson-on-mechanical-sympathy/) and is originally (in software) from [here](https://mechanical-sympathy.blogspot.com/2011/07/why-mechanical-sympathy.html).

Reply:
I went with an array of optional tables, so we don't lose the ability to detect missing tables